File: src/main/java/wraith/fabricaeexnihilo/modules/base/BaseBlockEntity.java
package wraith.fabricaeexnihilo.modules.base;

import net.minecraft.block.BlockState;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.network.Packet;
import net.minecraft.network.listener.ClientPlayPacketListener;
import net.minecraft.network.packet.s2c.play.BlockEntityUpdateS2CPacket;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.math.BlockPos;

public abstract class BaseBlockEntity extends BlockEntity {

    public BaseBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState state) {
        super(type, pos, state);
    }

    public void updateClient() {
        if (this.world != null && this.world instanceof ServerWorld serverWorld) {
            var chunkManager = serverWorld.getChunkManager();
            if (chunkManager != null) {
                chunkManager.markForUpdate(pos);
            }
        }
    }

    public void markDirtyClient() {
        this.markDirty();
        this.updateClient();
    }

    @Override
    public Packet<ClientPlayPacketListener> toUpdatePacket() {
        return BlockEntityUpdateS2CPacket.create(this);
    }

    @Override
    public NbtCompound toInitialChunkDataNbt() {
        return createNbt();
    }
}

File: src/main/kotlin/exnihilofabrico/modules/base/BaseBlockEntity.kt
package exnihilofabrico.modules.base

import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.server.world.ServerWorld

abstract class BaseBlockEntity(type: BlockEntityType<*>): BlockEntity(type) {
    fun updateClient() {
        (this.world as? ServerWorld)?.getChunkManager()?.markForUpdate(pos)
    }

    fun markDirtyClient() {
        this.markDirty()
        this.updateClient()
    }
}